### PR TITLE
for review

### PR DESCRIPTION
**Pull request by abhishek-rawat233**

### CODE REVIEW

File: app/controllers/reports_controller.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+class ReportsController < ApplicationController
+  include ReportConcerns
+  before_action :set_from_and_to
+
+  def index(from = @from, to = @to)
```
Comment (himanshumishra31) [inline app/controllers/reports_controller.rb]: Where are you fetching params date?

File: app/controllers/reports_controller.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+    @orders = Order.where(created_at: from..to)
```
Comment (himanshumishra31) [inline app/controllers/reports_controller.rb]: Create a scope for this. Pass values from param in that scope. Assign default values to scope in values not passed.

File: app/controllers/sessions_controller.rb [hunk 3]
```diff
@@ -6,8 +6,14 @@ def new
   def create
     user = User.find_by(name: params[:name])
     if user.try(:authenticate, params[:password])
-      session[:user_id] = user.id
-      redirect_to admin_url
+      if user.role == 'admin'
```
Comment (himanshumishra31) [inline app/controllers/sessions_controller.rb]: Use helper

File: app/controllers/sessions_controller.rb [hunk 4]
```diff
@@ -6,8 +6,14 @@ def new
...
+        session[:user_id] = user.id
+        redirect_to reports_path
+        # redirect_to admin_url
+      else
+        session[:user_id] = user.id
```
Comment (himanshumishra31) [inline app/controllers/sessions_controller.rb]: DRY

File: app/controllers/store_controller.rb [hunk 5]
```diff
@@ -9,4 +9,17 @@ def index
       @products = Product.order(:title)
     end
   end
+
+  def update_ratings
+    rating = Rating.where(user_id: @current_user, product_id: params[:product_id].to_i)
```
Comment (himanshumishra31) [inline app/controllers/store_controller.rb]: Before action

File: app/controllers/store_controller.rb [hunk 6]
```diff
@@ -9,4 +9,17 @@ def index
...
+    if rating.present?
+      rating.update_all(rating: params[:product_rating].to_f)
```
Comment (himanshumishra31) [inline app/controllers/store_controller.rb]: update_all for single object??

Comment (himanshumishra31) [inline app/controllers/store_controller.rb]: move it to model

File: app/controllers/store_controller.rb [hunk 7]
```diff
@@ -9,4 +9,17 @@ def index
...
+    else
+      Rating.create({rating: params[:product_rating].to_f,
+                    user_id: @current_user.id,
+                    product_id: params[:product_id].to_i})
+    end
+    product = Product.find(params[:product_id].to_i)
```
Comment (himanshumishra31) [inline app/controllers/store_controller.rb]: before_Action

File: app/javascript/packs/product_rating.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
+function update_ratings(e) {
```
Comment (himanshumishra31) [inline app/javascript/packs/product_rating.js]: Oops

File: app/controllers/admin_controller.rb [hunk 9]
```diff
@@ -1,5 +1,33 @@
...
+    if @category.present?
+      @products = @category.get_all_products
+    else
+      redirect_to categories_path, notice: "no such id present" unless @category.present?
+    end
+  end
+
+  private
+    def select_category
+      @category = Category.find_by(id: params[:id])
```
Comment (himanshumishra31) [inline app/controllers/admin_controller.rb]: Study joins includes

File: app/helpers/admin_helper.rb [hunk 10]
```diff
@@ -1,2 +1,5 @@
 module AdminHelper
+  def get_category_parent(category)
+    category.parent ? category.parent.name : 'no parent'
```
Comment (himanshumishra31) [inline app/helpers/admin_helper.rb]: category.parent.try(:name) || 'no parent'

File: app/models/category.rb [hunk 11]
```diff
@@ -13,6 +12,10 @@ class Category < ApplicationRecord
   before_destroy :sub_categories_has_products
   before_save :parent_cannot_be_subcategory, if: :parent_id?
 
+  def get_all_products
+    products + products_of_sub_categories
```
Comment (himanshumishra31) [inline app/models/category.rb]: Study eagerloading

File: app/models/product.rb [hunk 12]
```diff
@@ -46,6 +58,16 @@ def self.product_titles_in_line_items
                                   message: 'enter atleast 3 words separated with hypen(-) and no space and special characters allowed' }
   validates :permalink, uniqueness: true
 
+  def average_rating
+    total_rating = 0
+    total_users_rated = ratings.count
+    ratings.each do |rating|
```
Comment (himanshumishra31) [inline app/models/product.rb]: Use basic ruby methods

File: app/models/rating.rb [hunk 13]
```diff
@@ -0,0 +1,6 @@
+class Rating < ApplicationRecord
+  belongs_to :user
+  belongs_to :product
+
+  validates_uniqueness_of :user_id, scope: :product_id
```
Comment (himanshumishra31) [inline app/models/rating.rb]: Add validation for rating can't be negative

Comment (himanshumishra31) [reply app/models/rating.rb]: Don't use this syntax.

File: app/models/user.rb [hunk 14]
```diff
@@ -5,6 +5,8 @@ class User < ApplicationRecord
   validates :name, presence: true, uniqueness: true
   has_secure_password
   validates :email, uniqueness: true, format: {with: EMAIL_VALIDATOR, message: 'not valid' }
+  has_many :ratings
```
Comment (himanshumishra31) [inline app/models/user.rb]: Can't delete user if rating is present.

Comment (himanshumishra31) [reply app/models/user.rb]: Create an association to fetch all products which has been rated by the user.

File: app/controllers/reports_controller.rb (hunk 2 above)
Comment (abhishek-rawat233) [reply app/controllers/reports_controller.rb]: problem

File: app/models/product.rb (hunk 12 above)
Comment (abhishek-rawat233) [reply app/models/product.rb]: doubt

**Pull request closed without merging**